pkg/fpga/bitstream/bitstream.go

```diff
@@ -0,0 +1,169 @@
+// Copyright 2019 Intel Corporation. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bitstream
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	utilsexec "k8s.io/utils/exec"
+
+	"github.com/intel/intel-device-plugins-for-kubernetes/pkg/fpga/device"
+	"github.com/intel/intel-device-plugins-for-kubernetes/pkg/fpga/gbs"
+	"github.com/pkg/errors"
+)
+
+const (
+	fpgaconf        = "/opt/intel/fpga-sw/opae/fpgaconf-wrapper"
+	pciAddressRegex = `^[[:xdigit:]]{4}:([[:xdigit:]]{2}):([[:xdigit:]]{2})\.([[:xdigit:]])$`
+)
+
+var (
+	pciAddressRE = regexp.MustCompile(pciAddressRegex)
+)
+
+// FPGABitstream defines common interface for OPAE and OpenCL bitstreams
+type FPGABitstream interface {
+	Init() error
+	Validate(region, afu string) error
+	Program(fme *device.FPGADevice, execer utilsexec.Interface) error
+}
+
+// OPAEBitstream defines structure of .gbs file
+type OPAEBitstream struct {
+	Path   string
+	Region string
+	AFU    string
+}
+
+// Init gets Region and AFU from .gbs file
+func (bitstream *OPAEBitstream) Init() error {
+	gbs, err := gbs.Open(bitstream.Path)
+	if err != nil {
+		return errors.Wrapf(err, "%s: can't get bitstream info", bitstream.Path)
+	}
+
+	if len(gbs.Metadata.AfuImage.AcceleratorClusters) != 1 {
+		return errors.Errorf("%s: 'accelerator-clusters' field not found", bitstream.Path)
+	}
+
+	bitstream.Region = device.CanonizeID(gbs.Metadata.AfuImage.InterfaceUUID)
+	bitstream.AFU = device.CanonizeID(gbs.Metadata.AfuImage.AcceleratorClusters[0].AcceleratorTypeUUID)
```

Comment:
I have nowadays functions for that: `InterfaceUUID()` and `AcceleratorTypeUUID()`.
`AcceleratorTypeUUID()` also will return empty string in case AcceleratorCluster != 1.